ARCs/arc-0004.md

```diff
@@ -0,0 +1,527 @@
+* If `x` is an `N`-bit unsigned integer, `uint<N>`:
+    * `enc(x)` is the `N`-bit big-endian encoding of `x`.
+* If `x` is an `N`-bit unsigned fixed-point decimal number with precision `M`, `ufixed<N>x<M>`:
+    * `enc(x) = enc(x * 10^M)`, where `x * 10^M` is interpreted as a `uint<N>`.
+* If `x` is a boolean value `bool`:
+    * `enc(x)` is a single byte whose **most significant bit** is either 1 or 0, if `x` is true or false respectively. All other bits are 0. Note: this means that a value of true will be encoded as `0x80` (`10000000` in binary) and a value of false will be encoded as `0x00`. This is in contrast to most other encoding schemes, where a value of true is encoded as `0x01`.
+
+Since `string` and `address` are aliases for `byte[]` and `byte[32]`
+respectively (and `byte` is an alias for `uint8`), the rules for
+encoding these types are covered above.
+
+### "Foreign" Types
+
+Three special types are supported _only_ as the type of an
+argument. They may not be used in arrays or tuples, as their encoding
+is unusual.
+
+* `account` Represents an Algorand account.
+* `asset` Represents an Algorand Standard Asset (ASA)
+* `application` Represents an Algorand Smart Contract
+
+Some AVM opcodes require specific values to be placed in the "foreign
+arrays" of the application call transaction. These three types allow
+methods to describe these requirements. To encode method calls that
+have these types as arguments, the value in question is placed in the
+Accounts (`apat`), Foreign Assets (`apas`), or Foreign Apps (`apfa`)
+arrays, and a uint8 referring to that value is encoded in the normal
+location for the value. For the Accounts and Applications array, a 0
+implicitly refers to the Sender and the called application
+respectively. Therefore, the passed values begin at 1.
+
+When passing addresses, ASAs, or apps that are _not_ required to be
+accessed by such opcodes, ARC-4 Contracts **SHOULD** use the base
+types for passing these types: `address` and `uint64`.
+
+
+### Opt In, Close Out, and Clear State.
+
+An ARC-4 app **SHOULD** allow accounts to opt-in and close-out, using
+an application call transaction with no arguments an OnCompletion
+equal to OptIn (1) or CloseOut (2). Clients **MAY** determine that
+opt-in is unnecessary by querying the blockchain for the app's
+schema. Opt-in is unnecessary if the app has no local schema.  However,
+even if a local schema is present, some methods may be callable
+without opt-in. Clients **MAY** attempt method calls without opt-in
+and use the failure response to learn that opt-in is required.
+
+Clients **MAY** display the minimum balance increment that such an
+opt-in would require.  If an opt-in or close-out contains arguments,
+an ARC-4 app **SHOULD** interpret the arguments as an invocation as
+described in the rest of this document, to perform along with the
+opt-in or close-out.
+
+> Should the ABI allow contracts to specify which methods require
+> opt-in?
```

Comment:
Should this question be removed before merging?

Reply:
The intent is to leave that it to draw out comments while it's still considered to be in draft.